File: swift/source/abstract/Tracer.swift
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

final class Tracer {
    typealias SQLTracer = (String) -> Void // SQL
    private var sqlTracer: SQLTracer? = nil {
        didSet {
            setup()
        }
    }

    typealias PerformanceTracer = ([String: Int], Int64, Any?) -> Void // (SQL, count), cost, userInfo
    private var performanceTracer: PerformanceTracer? = nil {
        didSet {
            setup()
        }
    }
    var shouldAggregation = false
    private var footprint: [String: Int] = [:]
    private var cost: Int64 = 0

    private let handle: SQLite3
    var userInfo: Any?

    init(with handle: SQLite3) {
        self.handle = handle
    }

    deinit {
        reportPerformance()
    }

    struct TraceType: OptionSet {
        let rawValue: UInt32

        static let stmt = TraceType(rawValue: UInt32(SQLITE_TRACE_STMT))
        static let profile = TraceType(rawValue: UInt32(SQLITE_TRACE_PROFILE))
    }

    private var type: TraceType {
        var type = TraceType()
        if sqlTracer != nil {
            type.insert(.stmt)
        }
        if performanceTracer != nil {
            type.insert(.profile)
        }
        return type
    }

    private func setup() {
        let traceType = self.type
        guard !traceType.isEmpty else {
            sqlite3_trace_v2(handle, 0, nil, nil)
            return
        }
        sqlite3_trace_v2(handle, traceType.rawValue, { (flag, context, stmtPointer, costPointer) -> Int32 in
            let traceType = TraceType(rawValue: flag)
            let pointer = Unmanaged<Tracer>.fromOpaque(context!)
            let tracer = pointer.takeUnretainedValue()

            let stmt = OpaquePointer(stmtPointer)
            guard let csql = sqlite3_sql(stmt) else {
                return SQLITE_MISUSE
            }
            let sql = String(cString: csql)
            switch traceType {
            case .stmt:
                tracer.report(sql: sql)
            case .profile:
                let cost: Int64 = UnsafePointer<Int64>(OpaquePointer(costPointer)!).pointee

                //report last track 
                if !tracer.shouldAggregation {
                    tracer.reportPerformance()
                }

                tracer.recordPerformance(sql: sql, cost: cost)
            default: break
            }
            return SQLITE_OK
        }, Unmanaged.passUnretained(self).toOpaque())
    }

    private func report(sql: String) {
        guard let sqlTracer = self.sqlTracer else {
            return
        }
        sqlTracer(sql)
    }

    private func reportPerformance() {
        guard !footprint.isEmpty else {
            return
        }
        guard let performanceTracer = self.performanceTracer else {
            return
        }
        performanceTracer(footprint, cost, userInfo)
        footprint.removeAll()
        cost = 0
    }

    private func recordPerformance(sql: String, cost: Int64) {
        self.footprint[sql] = self.footprint[sql] ?? 1
        self.cost = cost
    }

    func trace(sql sqlTracer: @escaping SQLTracer) {
        self.sqlTracer = sqlTracer
    }

    func track(performance performanceTracer: @escaping PerformanceTracer) {
        self.performanceTracer = performanceTracer
    }
}

File: objc/source/abstract/Tracer.cpp
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <WCDB/Tracer.hpp>
#include <sqlite3.h>

namespace WCDB {

Tracer::Footprint::Footprint(const std::string &sql_) : sql(sql_), frequency(1)
{
}

int Tracer::TraceCallback(unsigned int flag, void *M, void *P, void *X)
{
    Tracer *tracer = (Tracer *) M;
    sqlite3_stmt *stmt = (sqlite3_stmt *) P;
    switch (flag) {
        case SQLITE_TRACE_STMT: {
            const char *sql = sqlite3_sql(stmt);
            if (sql) {
                tracer->reportSQL(sql);
            }
        } break;
        case SQLITE_TRACE_PROFILE: {
            sqlite3_int64 *cost = (sqlite3_int64 *) X;

            sqlite3 *db = sqlite3_db_handle(stmt);

            const char *sql = sqlite3_sql(stmt);
            if (sql) {
                tracer->addPerformanceTrace(sql, *cost);
            }

            //report per transaction
            if (sqlite3_get_autocommit(db)) {
                tracer->reportPerformance();
            }
        } break;
        default:
            break;
    }
    return SQLITE_OK;
}

void Tracer::setupTrace(void *handle)
{
    unsigned flag = 0;
    if (m_sqlTrace) {
        flag |= SQLITE_TRACE_STMT;
    }
    if (m_performanceTrace) {
        flag |= SQLITE_TRACE_PROFILE;
    }
    if (flag != 0) {
        sqlite3_trace_v2((sqlite3 *) handle, flag, Tracer::TraceCallback, this);
    } else {
        sqlite3_trace_v2((sqlite3 *) handle, 0, nullptr, nullptr);
    }
}

void Tracer::traceSQL(void *handle, const SQLTraceCallback &trace)
{
    m_sqlTrace = trace;
    setupTrace(handle);
}

void Tracer::tracePerformance(void *handle,
                              const PerformanceTraceCallback &trace)
{
    m_performanceTrace = trace;
    setupTrace(handle);
}

void Tracer::addPerformanceTrace(const std::string &sql, const int64_t &cost)
{
    if (!m_footprints.empty()) {
        auto &footprint = m_footprints.front();
        if (footprint.sql == sql) {
            ++footprint.frequency;
            return;
        }
    }
    m_footprints.push_back(Footprint(sql));
    m_cost += cost;
}

void Tracer::reportPerformance()
{
    if (!m_footprints.empty()) {
        m_performanceTrace(m_footprints, m_cost);
        m_footprints.clear();
        m_cost = 0;
    }
}

void Tracer::reportSQL(const std::string &sql)
{
    if (m_sqlTrace) {
        m_sqlTrace(sql);
    }
}

} //namespace WCDB
